### DISCUSSION TITLE
High MAPE

### DISCUSSION THREAD

**Discussion by huydinhtran** [accel-sim/accel-sim-framework#182] [Q&A]
Hi,

So I'm following the "[AccelWattch MICRO'21 Artifact Appendix Manual](https://github.com/accel-sim/accel-sim-framework/blob/release/AccelWattch.md)" using a GTX 1050Ti. The simulated power results with SASS and PTX I got were not that accurate compared to the real power results. I got around 60% MAPE on average while in the paper they stated that they got a 11±3.8% MAPE with the Pascal TITAN X. What are some potential reasons for this inaccuracy?

Thanks!

**Comment by JRPan** [comment]
Can you check how the performance counter (cache, memory, cycle, etc.) looks? Are they correlating well with the hw?

**Reply by huydinhtran** [reply]
Is that from running the "[profile_validation_perf.sh](https://github.com/accel-sim/accel-sim-framework/blob/release/AccelWattch.md#collecting-hardware-performance-counter-information-for-validation-kernels)"? I ran it but it gave an error below. 

```
Traceback (most recent call last):
  File "/home/htran197/myStuff/accel-sim-framework/gpu-simulator/../util/hw_stats/run_hw.py", line 123, in <module>
    exec_path = common.file_option_test(os.path.join(edir, exe),"",this_directory)
  File "/home/htran197/myStuff/accel-sim-framework/util/hw_stats/../job_launching/common.py", line 160, in file_option_test
    raise PathMissing("Error - cannot open file {0}".format(name))
common.PathMissing: Error - cannot open file /home/htran197/myStuff/accel-sim-framework/gpu-app-collection/src/..//bin/11.0/release/backprop-rodinia-2.0-ft
Done
```

**Reply by JRPan** [reply]
I meant running Accel-Sim only. Without the power model. 
Run `./util/job_launching/run_simulations.py -B rodinia_2.0-ft -C QV100-SASS -T ./hw_run/traces/device-<device-num>/<cuda-version>/ -N myTest`, then collect HW stats `./util/tracer_nvbit/run_hw_trace.py -B rodinia_2.0-ft -D <gpu-device-num-to-run-on>`, and last collect results and correlate
`./util/job_launching/get_stats.py -k -K -R -N all-apps | tee all-apps.csv`
`./util/plotting/plot-correlation.py -c all-apps.csv -H ./hw_run/QUADRO-V100/device-0/9.1/`

Or you can check the results manually. The steps are documented here
https://github.com/accel-sim/accel-sim-framework

The power model depends on Accel-Sim's counters. So to get accurate power, you have to get accurate counters first.

**Reply by huydinhtran** [reply]
I was trying to run `./util/tracer_nvbit/run_hw_trace.py -B rodinia_2.0-ft -D 0 `. It gave me an error of `GPGPU-Sim ** ERROR: Cannot open config file 'gpgpusim.config'`. What are the causes for this?

**Reply by JRPan** [reply]
Start a new terminal to clear environmental variables. You sourced the setup_envrionment script, so the app is looking for gpgpu-sim instead of the real GPU.

**Reply by huydinhtran** [reply]
At `./util/job_launching/get_stats.py -k -K -R -N all-apps | tee all-apps.csv` I received `Could not find logfiles for job with the name "all-apps"`

**Comment by JRPan** [comment]
The name is the what you set when running run_simulationr.py
For example, `./util/job_launching/run_simulations.py -B rodinia_2.0-ft -C QV100-SASS -T ./hw_run/traces/device-<device-num>/<cuda-version>/ -N myTest` the name is myTest. You can ignore the `-N all-apps` option. It will get the latest stat if not specified.

**Reply by JRPan** [reply]
Also, I'm not sure if this applies to your branch or not. Check this file: `util/hw_stats/run_hw.py` to check this line https://github.com/accel-sim/accel-sim-framework/blob/49e02f4d36cad0a85491d11d437249f29e60725d/util/hw_stats/run_hw.py#L71.

You probably need to add the option `-C other_stats` when using run_hw.py to collect HW stats.

Thanks

**Reply by huydinhtran** [reply]
For `./util/job_launching/get_stats.py -k -K -R -N myTest | tee myTest.csv`, I believe it ran fine since it printed a lot of text and some numbers.

However, at `./util/plotting/plot-correlation.py -c all-apps.csv -H ./hw_run/traces/device-0/11.0/`, it said the output is in `correl-html` directory but it is not there.

**Reply by JRPan** [reply]
I meant when collecting HW stats with `run_hw`. I remember by default, it only collects some basic stats. 

You are running 1050Ti right? Which we don't have. To get the plots you have to change here, I think
https://github.com/accel-sim/accel-sim-framework/blob/release/util/plotting/correl_mappings.py#L5-L15

Left should be the gpgpu-sim config name, right should be the name reported by the Nvidia profiler. For example, here is from a profiler output of a V100. You should find these files under `./hw_run`
You can see the device name is `Tesla V100-SXM2-32GB` and this is what you should put in the map. Or you can run `nvidia-smi -q` and look for the field `Product Name`. The name should be the same
```
==78712== Profiling result:
"Device","Context","Stream","Kernel","Correlation_ID","inst_per_warp","branch_efficiency","warp_execution_efficiency","warp_nonpred_execution_efficiency","inst_replay_overhead","shared_load_transactions_per_request","shared_store_transactions_per_request","local_load_transactions_per_request","local_store_transactions_per_request","gld_transactions_per_request","gst_transactions_per_request","shared_store_transactions","shared_load_transactions","local_load_transactions","local_store_transactions","gld_transactions","gst_transactions","sysmem_read_transactions","sysmem_write_transactions","l2_read_transactions","l2_write_transactions","dram_read_transactions","dram_write_transactions","global_hit_rate","local_hit_rate","gld_requested_throughput","gst_requested_throughput","gld_throughput","gst_throughput","local_memory_overhead","tex_cache_hit_rate","l2_tex_read_hit_rate","l2_tex_write_hit_rate","dram_read_throughput","dram_write_throughput","tex_cache_throughput","l2_tex_read_throughput","l2_tex_write_throughput","l2_read_throughput","l2_write_throughput","sysmem_read_throughput","sysmem_write_throughput","local_load_throughput","local_store_throughput","shared_load_throughput","shared_store_throughput","gld_efficiency","gst_efficiency","tex_cache_transactions","flop_count_dp","flop_count_dp_add","flop_count_dp_fma","flop_count_dp_mul","flop_count_sp","flop_count_sp_add","flop_count_sp_fma","flop_count_sp_mul","flop_count_sp_special","inst_executed","inst_issued","dram_utilization","sysmem_utilization","stall_inst_fetch","stall_exec_dependency","stall_memory_dependency","stall_texture","stall_sync","stall_other","stall_constant_memory_dependency","stall_pipe_busy","shared_efficiency","inst_fp_32","inst_fp_64","inst_integer","inst_bit_convert","inst_control","inst_compute_ld_st","inst_misc","inst_inter_thread_communication","issue_slots","cf_issued","cf_executed","ldst_issued","ldst_executed","atomic_transactions","atomic_transactions_per_request","l2_atomic_throughput","l2_atomic_transactions","l2_tex_read_transactions","stall_memory_throttle","stall_not_selected","l2_tex_write_transactions","flop_count_hp","flop_count_hp_add","flop_count_hp_mul","flop_count_hp_fma","inst_fp_16","ipc","issued_ipc","issue_slot_utilization","sm_efficiency","achieved_occupancy","eligible_warps_per_cycle","shared_utilization","l2_utilization","tex_utilization","ldst_fu_utilization","cf_fu_utilization","tex_fu_utilization","special_fu_utilization","half_precision_fu_utilization","single_precision_fu_utilization","double_precision_fu_utilization","flop_hp_efficiency","flop_sp_efficiency","flop_dp_efficiency","sysmem_read_utilization","sysmem_write_utilization","stall_sleeping","pcie_total_data_transmitted","pcie_total_data_received","inst_executed_global_loads","inst_executed_local_loads","inst_executed_shared_loads","inst_executed_surface_loads","inst_executed_global_stores","inst_executed_local_stores","inst_executed_shared_stores","inst_executed_surface_stores","inst_executed_global_atomics","inst_executed_global_reductions","inst_executed_surface_atomics","inst_executed_surface_reductions","inst_executed_shared_atomics","inst_executed_tex_ops","dram_read_bytes","dram_write_bytes","global_load_requests","local_load_requests","surface_load_requests","global_store_requests","local_store_requests","surface_store_requests","global_atomic_requests","global_reduction_requests","surface_atomic_requests","surface_reduction_requests","l2_global_load_bytes","l2_local_load_bytes","l2_surface_load_bytes","l2_global_atomic_store_bytes","l2_local_global_store_bytes","l2_surface_store_bytes","sysmem_read_bytes","sysmem_write_bytes","l2_tex_hit_rate","texture_load_requests","tensor_precision_fu_utilization"
,,,,,,%,%,%,,,,,,,,,,,,,,,,,,,,%,%,GB/s,GB/s,GB/s,GB/s,%,%,%,%,GB/s,GB/s,GB/s,GB/s,GB/s,GB/s,GB/s,B/s,MB/s,B/s,B/s,GB/s,GB/s,%,%,,,,,,,,,,,,,,,%,%,%,%,%,%,%,%,%,,,,,,,,,,,,,,,,B/s,,,%,%,,,,,,,,,%,%,,,,,,,,,,,,,%,%,%,,,%,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,%,,
"Tesla V100-SXM2-32GB (0)","1","7","_Z22bpnn_layerforward_CUDAPfS_S_S_ii",1,242.000000,66.666667,77.530992,62.835744,0.002851,1.178297,1.004676,0.000000,0.000000,2.946976,3.000000,160491,299231,0,0,193133,196608,0,5,156316,196652,69662,142600,61.987576,0.000000,201.231061,201.231061,279.069669,284.090909,57.940577,12.245396,66.015188,100.000000,100.658879,206.051451,2239.013441,224.902413,284.090909,225.870537,284.154487,0.000000,7.398200,0.000000,0.000000,1729.508602,927.612998,72.107822,70.833333,387383,0,0,0,0,2031616,983040,0,1048576,4194304,7929856,4502015,"Mid (4)","Low (1)",1.039380,19.151213,30.497507,0.160544,16.733647,3.598171,1.096262,12.954813,49.894501,6225920,0,57409536,8388608,1048576,10551296,19922944,0,4502015,98304,98304,1114112,1114112,0,0.000000,0.000000,0,155646,1.048981,13.719483,196608,0,0,0,0,0,2.517086,2.336741,58.418535,77.976086,0.901912,8.959438,"Low (1)","Low (1)","Low (2)","Mid (6)","Low (1)","Idle (0)","Mid (5)","Idle (0)","Mid (4)","Idle (0)",0.000000,0.707499,0.000000,"Idle (0)","Low (1)",0.000000,0,0,65536,0,253952,0,65536,0,159744,0,0,0,0,0,0,0,2229184,4563200,65536,0,0,65536,0,0,0,0,0,0,4607072,0,0,0,6291456,0,0,160,84.983563,0,"Idle (0)"
"Tesla V100-SXM2-32GB (0)","1","7","_Z24bpnn_adjust_weights_cudaPfiS_iS_S_",2,96.002319,99.996948,99.998792,96.938924,0.010096,0.000000,0.000000,0.000000,0.000000,3.244236,4.999939,0,0,0,0,744166,327686,0,5,303523,327713,281444,191148,74.021507,0.000000,586.908733,284.560889,807.790022,355.702197,55.775393,42.549071,17.929177,100.000000,305.506641,207.490596,1315.652857,328.884991,355.702197,329.473331,355.731505,0.000000,5.557745,0.000000,0.000000,0.000000,0.000000,72.656101,79.999756,303007,9437296,1048592,2097184,4194336,0,0,0,0,0,3145804,1456372,"High (7)","Low (1)",0.476488,4.594851,74.449890,0.000310,11.299315,0.400518,0.739144,2.936828,0.000000,0,7340112,9437232,9437296,1048576,9437296,8388608,0,1456372,65537,65537,425991,425991,0,0.000000,0.000000,0,302981,4.079661,1.022995,327686,0,0,0,0,0,0.582350,0.597418,14.935450,81.859934,0.910665,1.179472,"Idle (0)","Low (1)","Low (2)","Low (2)","Low (1)","Idle (0)","Low (3)","Idle (0)","Low (1)","Low (1)",0.000000,0.000000,1.792181,"Idle (0)","Low (1)",0.000000,512,0,229381,0,0,0,65538,0,0,0,0,0,0,0,0,0,9006208,6116736,229381,0,0,65538,0,0,0,0,0,0,8867840,0,0,0,10485952,0,0,160,60.572061,0,"Idle (0)"
```
and
```
pan251@tgrogers-gpu01:~$ nvidia-smi -q | grep 'Product Name'
    Product Name                          : Tesla V100-SXM2-32GB
```

Let me know if this works

**Reply by huydinhtran** [reply]
I'm quite confused here since we didn't touch the `run_hw.py` file before. Do I run `./util/hw_stats/run_hw.py -B rodinia_2.0-ft -C other_stats` after the steps that you mentioned?

**Reply by huydinhtran** [reply]
So I ran `./util/hw_stats/run_hw.py -B rodinia_2.0-ft -C other_stats` and this is the output I got which doesn't look similar to yours. What are the potential causes for this?

`==32411== Profiling result:
"Start","Duration","Grid X","Grid Y","Grid Z","Block X","Block Y","Block Z","Registers Per Thread","Static SMem","Dynamic SMem","Size","Throughput","SrcMemType","DstMemType","Device","Context","Stream","Name","Correlation_ID"
us,us,,,,,,,,KB,B,KB,GB/s,,,,,,,
224831.753000,2.816000,,,,,,,,,,16.003906,5.419927,"Pageable","Device","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy HtoD]",114
224885.387000,45.058000,,,,,,,,,,272.066406,5.758417,"Pageable","Device","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy HtoD]",115
224937.806000,19.584000,1,256,1,16,16,1,20,1.062500,0,,,,,"NVIDIA GeForce GTX 1050 Ti (0)","1","7","_Z22bpnn_layerforward_CUDAPfS_S_S_ii",116
224970.927000,2.368000,,,,,,,,,,16.000000,6.443745,"Device","Pageable","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy DtoH]",119
225018.353000,0.864000,,,,,,,,,,0.066406,0.073299,"Pageable","Device","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy HtoD]",122
225051.602000,44.642000,,,,,,,,,,272.066406,5.812077,"Pageable","Device","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy HtoD]",123
225121.973000,44.802000,,,,,,,,,,272.066406,5.791321,"Pageable","Device","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy HtoD]",124
225169.847000,38.817000,1,256,1,16,16,1,20,0.000000,0,,,,,"NVIDIA GeForce GTX 1050 Ti (0)","1","7","_Z24bpnn_adjust_weights_cudaPfiS_iS_S_",125
225209.816000,2.304000,,,,,,,,,,16.003906,6.624355,"Device","Pageable","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy DtoH]",126
225223.993000,37.345000,,,,,,,,,,272.066406,6.947724,"Device","Pageable","NVIDIA GeForce GTX 1050 Ti (0)","1","7","[CUDA memcpy DtoH]",127
`

**Reply by JRPan** [reply]
Try`./util/hw_stats/run_hw.py -B rodinia_2.0-ft --nsight_profiler --disable_nvprof -C other_stats` 
The steps are detailed in the readme of Accel-Sim.

**Reply by huydinhtran** [reply]
I'm not too sure what I did wrong here since I tried to follow the readme and couldn't get what you had. To make it simple, I just changed the gpgpu-sim config file of TITANX to 1050Ti and keep the same name. I also changed the file `./util/plotting/correl_mappings.py` to `"TITANX" : set("NVIDIA GeForce GTX 1050 Ti")`
Here are the steps that I did:

1. `./util/tracer_nvbit/run_hw_trace.py -B rodinia_2.0-ft -D 0`
2. `./util/job_launching/run_simulations.py -B rodinia_2.0-ft -C TITANX-SASS -T ./hw_run/traces/device-0/11.0/ -N myTest`
3. `./util/hw_stats/run_hw.py -B rodinia_2.0-ft -C other_stats`  (`--nsight_profiler --disable_nvprof` failed)

However, the output file in `./hw_run/` is still similar to what I had above. I might have run the steps wrong in some way that I'm not aware of.

**Reply by MWHYNOT** [reply]
@JRPan  Hello, JRPan, I wonder is the command  `. /util/job_launching/run_simulations.py -B rodinia_2.0-ft -C QV100-SASS -T . /hw_run/traces/device-<device-num>/<cuda-version>/ -N myTest`  running all the rodinia_2.0-ft benchmark suite using the simulator ?
and what is the purpose of `-N myTest`?

When I use the command `. /util/job_launching/run_simulations.py -B rodinia_2.0-ft -C QV100-SASS -T . /hw_run/traces/device-<device-num>/<cuda-version>/ -N myTest`, it always appears error:
`Cannot resolve default server host 'torqueserver' - check server_name file.
qsub: cannot connect to server torqueserver (errno=15008) Access from host not allowed, or unknown host
Job  queued (backprop-rodinia-2.0-ft-4096___data_result_4096_txt QV100-SASS)`

So I have to run the benchmark using the command  `./justrun` in the /accel-sim-framework/sim_run_11.2/bfs-rodinia-2.0-ft/__data_graph4096_txt___data_graph4096_result_txt/QV100-SASS/

I would really appreciate it if you could give me some advice

**Comment by JRPan** [comment]
You already got the simulation results. `run_hw_trace`is to get the traces, and run simulation with `run_simulations`. `run_hw` is not depending on any of those. So you don't need to worry about anything else other than the `run_hw`. 
What is the fail msg? Do you have nsight compute installed? It should come with the CUDA package.

**Reply by huydinhtran** [reply]
I installed CUDA as how it is in the instruction. 
When running with `--nsight_profiler --disable_nvprof`, I got the error here.
```
==ERROR== Error: ERR_NVGPUCTRPERM - The user does not have permission to access NVIDIA GPU Performance Counters on the target device 0. For instructions on enabling permissions and to get more information see https://developer.nvidia.com/ERR_NVGPUCTRPERM
==PROF== Disconnected from process 3803
==ERROR== An error occurred while trying to profile.
==WARNING== No kernels were profiled.
```


I'm very confused here. So I followed instructions exactly from the `Artifact Manual` using the old `release-accelwattch` branch. After receiving the SASS and PTX simulated results, I just need to run `./util/hw_stats/run_hw.py -B rodinia-3.1_validation,parboil_validation,cuda_samples_11.0_validation -C other_stats` (benchmarks are from that manual) right? So we don't run the steps mentioned at the beginning of the discussion of the `run_simulations.py` and `run_hw_trace.py` scripts?

**Reply by JRPan** [reply]
I asked you to run all `run_hw`, `run_simulations` and `run_hw_trace`. I don't remember the exact steps in the artifact. But I asked you to do this because you said you have a high error, and I want you to check if the performance counters match with the real HW. 

`run_simulations` is just launching the Accel-Sim to run simulations. If you have already done that, then no need to re-run the simulator. 
`run_hw_trace` is to collect the HW traces using your physical GPU. You cannot use traces collected on a different GPU generation (for example, you cannot use an Ampere trace collected on a 3070 to simulate a 1050Ti GPU). This is not needed for PTX mode.

Now, assuming your simulation is finished with the correct traces, you have to compare the simulation results with counters from the real GPU. To collect those data, you use `run_hw`. The error you are having is because you don't have correct permission. To resolve that, please follow the link from the output. 

After you have collected HW data, correlate HW with the simulator results. The error would probably be high. Then you need to update your config file, then see if the correlation is better. 
When the correlation improves, your MAPE should also improve. Because the power is calculated based on the counters from the Accel-Sim.
